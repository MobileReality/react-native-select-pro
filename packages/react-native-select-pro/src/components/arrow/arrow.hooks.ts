import { useSelectContext } from '../../context';
import { useAnimation } from '../../hooks';

export const useArrow = () => {
    const {
        isOpened,
        styles: mainStyles,
        animation,
        arrowImageProps,
        arrowContainerProps,
    } = useSelectContext();

    const { arrow: arrowStyles } = mainStyles?.select ?? {};
    const { icon: iconCustomStyles, container: containerCustomStyles } = arrowStyles ?? {};

    const rotateAnimation = useAnimation({ isOpened, animation });

    return {
        arrowImageProps,
        arrowContainerProps,
        rotateAnimation,
        iconCustomStyles,
        containerCustomStyles,
        isOpened,
    };
};
