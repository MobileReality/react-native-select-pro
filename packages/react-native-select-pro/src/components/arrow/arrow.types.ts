import type { ComponentProps } from 'react';

import type { OptionalToRequired } from '../../helpers';
import type { State } from '../../state/types';
import type { OnPressSelectControlType } from '../../types';
import type { Select } from '../select';

export type ArrowProps = OptionalToRequired<
    { onPressSelectControl: OnPressSelectControlType } & Pick<State, 'isOpened'> &
        Pick<
            ComponentProps<typeof Select>,
            | 'animation'
            | 'multiSelection'
            | 'disabled'
            | 'selectControlArrowImageStyle'
            | 'customSelectControlArrowIconSource'
        >
>;
