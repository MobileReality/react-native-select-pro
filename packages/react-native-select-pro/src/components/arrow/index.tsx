import type { ComponentProps } from 'react';
import React from 'react';
import type { ImageStyle } from 'react-native';
import { Animated, Image, Pressable, StyleSheet } from 'react-native';

import type { OptionalToRequired } from '../../helpers/types/optional-to-required';
import { useAnimation } from '../../hooks/use-animation';
import type { State } from '../../state/types';
import type { OnPressSelectControlType } from '../../types';
import type { Select } from '../select';

const arrowImage = require('./../../assets/icons/chevron-down.png');

type ArrowProps = OptionalToRequired<
    { onPressSelectControl: OnPressSelectControlType } & Pick<State, 'isOpened'> &
        Pick<
            ComponentProps<typeof Select>,
            | 'animation'
            | 'multiSelection'
            | 'disabled'
            | 'selectControlArrowImageStyle'
            | 'customSelectControlArrowIconSource'
        >
>;

export const Arrow = ({
    isOpened,
    disabled,
    animation,
    multiSelection,
    selectControlArrowImageStyle,
    customSelectControlArrowIconSource,
    onPressSelectControl,
}: ArrowProps) => {
    const rotateAnimation = useAnimation({ isOpened, animation });
    const accessibilityLabel = 'Arrow for opening dropdown';

    const resolveContent = () => {
        const arrowSource = customSelectControlArrowIconSource ?? arrowImage;

        if (rotateAnimation) {
            const rotate = rotateAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: ['0deg', '180deg'],
            });

            return (
                <Animated.Image
                    source={arrowSource}
                    style={[
                        styles.arrowIcon,
                        { transform: [{ rotate }] },
                        selectControlArrowImageStyle,
                    ]}
                />
            );
        }

        return (
            <Image
                source={arrowSource}
                style={[
                    styles.arrowIcon,
                    isOpened ? styles.arrowIconOpened : styles.arrowIconClosed,
                    selectControlArrowImageStyle,
                ]}
            />
        );
    };
    const arrowContent = resolveContent();

    if (!multiSelection) {
        return arrowContent;
    }

    return (
        <Pressable
            accessibilityLabel={accessibilityLabel}
            onPress={disabled ? undefined : onPressSelectControl}
        >
            {arrowContent}
        </Pressable>
    );
};

type Styles = {
    arrowIcon: ImageStyle;
    arrowIconOpened: ImageStyle;
    arrowIconClosed: ImageStyle;
};

const styles = StyleSheet.create<Styles>({
    arrowIcon: {
        width: 25,
        height: 25,
        zIndex: -1,
    },
    arrowIconOpened: {
        transform: [{ rotate: '180deg' }],
    },
    arrowIconClosed: {
        transform: [{ rotate: '0deg' }],
    },
});
