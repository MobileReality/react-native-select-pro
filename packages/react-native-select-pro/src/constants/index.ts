export * from './others';
export * from './portals';
export * from './styles';
