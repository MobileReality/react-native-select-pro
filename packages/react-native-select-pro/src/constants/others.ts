import { NativeModules, Platform } from 'react-native';

// based on: https://github.com/callstack/react-native-paper/blob/main/src/constants.tsx
const statusBarHeight =
    NativeModules.NativeUnimoduleProxy?.modulesConstants?.ExponentConstants?.statusBarHeight ?? 0;

export const APPROX_STATUSBAR_HEIGHT: number = Platform.select({
    android: statusBarHeight,
    ios: Platform.Version < 11 ? statusBarHeight : 0,
});
