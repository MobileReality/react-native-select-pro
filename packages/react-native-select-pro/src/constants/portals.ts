export enum Portals {
    Backdrop = 'backdrop',
    OptionsList = 'optionsList',
}
