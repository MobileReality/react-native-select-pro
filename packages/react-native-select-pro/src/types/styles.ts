import type { ImageStyle, StyleProp, TextStyle, ViewStyle } from 'react-native';

export type SelectStyles = {
    select?: SelectControlStyles;
    option?: OptionStyles;

    /**
     * Style for the options list container.
     */
    optionsList?: StyleProp<ViewStyle>;

    noOptions?: {
        /**
         *  Style for the no options container.
         */
        container?: StyleProp<ViewStyle>;

        /**
         *  Style for the no options text.
         */
        text?: StyleProp<TextStyle>;
    };
    sectionHeader?: SectionHeaderStyles;

    /**
     * Style for backdrop.
     */
    backdrop?: StyleProp<ViewStyle>;
};

export type SelectControlStyles = {
    /**
     * Style for the select.
     */
    container?: StyleProp<ViewStyle>;
    /**
     *  Style for the select text.
     */
    text?: StyleProp<TextStyle>;

    /**
     *  Style for the disabled select.
     */
    disabled?: StyleProp<ViewStyle>;

    /**
     *  Style for the buttons container the right side.
     */
    buttons?: StyleProp<ViewStyle>;

    multiSelectedOption?: {
        /**
         *  Style for the selected option in select control.
         */
        container?: StyleProp<ViewStyle>;

        /**
         *  Style for the selected option text in select control.
         */
        text?: StyleProp<TextStyle>;

        /**
         * Style for the pressed selected option in select control.
         */
        pressed?: StyleProp<ViewStyle>;
    };

    /**
     *  Style for the arrow.
     */
    arrow?: {
        /**
         * Style for the arrow container.
         */
        container?: StyleProp<ViewStyle>;

        /**
         *  Style for the arrow icon.
         */
        icon?: StyleProp<ImageStyle>;
    };

    clear?: {
        /**
         * Style for the clear option button.
         */
        container?: StyleProp<ViewStyle>;
        /**
         *  Style for the clear option icon.
         */
        icon?: StyleProp<ImageStyle>;
    };

    /**
     * Style for the custom left icon.
     */
    leftIcon?: StyleProp<ImageStyle>;
};

export type OptionStyles = {
    /**
     * Style for the single option.
     */
    container: StyleProp<ViewStyle>;

    /**
     * Style for the pressed single option.
     */
    pressed?: StyleProp<ViewStyle>;

    selected?: {
        /**
         * Style for the selected single option.
         */
        container: StyleProp<ViewStyle>;

        /**
         * Style for the selected single option text.
         */
        text?: StyleProp<TextStyle>;
    };

    /**
     * Style for the single option text.
     */
    text?: StyleProp<TextStyle>;
};

export type SectionHeaderStyles = {
    /**
     * Style for the section header.
     */
    container?: StyleProp<ViewStyle>;

    clear?: {
        /**
         * Style for the section header clear icon.
         */
        icon?: StyleProp<ImageStyle>;
    };

    /**
     * Style for the pressed section header.
     */
    pressed?: StyleProp<ViewStyle>;

    selected?: {
        /**
         * Style for section header when all section options are selected
         */
        container?: StyleProp<ViewStyle>;

        /**
         * Style for section header text when all section options are selected.
         */
        text?: StyleProp<TextStyle>;
    };

    /**
     * Style for the section header text.
     */
    text?: StyleProp<TextStyle>;
};
