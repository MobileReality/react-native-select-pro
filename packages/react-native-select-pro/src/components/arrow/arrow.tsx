import React, { useMemo } from 'react';
import type { ImageStyle } from 'react-native';
import { Animated, Image, StyleSheet, View } from 'react-native';

import { useArrow } from './arrow.hooks';

const arrowImage = require('./../../assets/icons/chevron-down.png');

export const Arrow = () => {
    const {
        arrowImageProps,
        arrowContainerProps,
        rotateAnimation,
        containerCustomStyles,
        iconCustomStyles,
        isOpened,
    } = useArrow();

    const renderImage = useMemo(() => {
        if (rotateAnimation) {
            const rotate = rotateAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: ['0deg', '180deg'],
            });

            return (
                <Animated.Image
                    source={arrowImage}
                    {...arrowImageProps}
                    style={[styles.arrowIcon, { transform: [{ rotate }] }, iconCustomStyles]}
                />
            );
        }

        return (
            <Image
                source={arrowImage}
                {...arrowImageProps}
                style={[styles.arrowIcon, isOpened && styles.arrowIconOpened, iconCustomStyles]}
            />
        );
    }, [iconCustomStyles, arrowImageProps, isOpened, rotateAnimation]);

    return (
        <View testID="Dropdown arrow" {...arrowContainerProps} style={containerCustomStyles}>
            {renderImage}
        </View>
    );
};

type Styles = {
    arrowIcon: ImageStyle;
    arrowIconOpened: ImageStyle;
};

const styles = StyleSheet.create<Styles>({
    arrowIcon: {
        width: 24,
        height: 24,
        zIndex: -1,
    },
    arrowIconOpened: {
        transform: [{ rotate: '180deg' }],
    },
});
