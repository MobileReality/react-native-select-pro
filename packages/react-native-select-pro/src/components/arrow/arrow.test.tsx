import React from 'react';

import { ANIMATION_DURATION } from '../../constants/styles';

import { Arrow } from './arrow';

describe('Arrow', () => {
    it('should generate Arrow snapshot', () => {
        const arrow = (
            <Arrow
                isOpened={true}
                disabled={false}
                animation={ANIMATION_DURATION}
                multiSelection={false}
                arrowIconStyles={{}}
                onPressSelectControl={() => {}}
            />
        );
        expect(arrow).toMatchSnapshot();
    });
});
