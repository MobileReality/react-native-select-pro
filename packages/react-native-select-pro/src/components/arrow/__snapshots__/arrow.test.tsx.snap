// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow should generate Arrow snapshot 1`] = `
<Arrow
  animation={200}
  arrowIconStyles={Object {}}
  disabled={false}
  isOpened={true}
  multiSelection={false}
  onPressSelectControl={[Function]}
/>
`;
